### PR TITLE
 Do not use iterators in find() and emplace() methods of hash tables.

### PR DESCRIPTION
**Pull request by akuzm**
Instead, these methods return a pointer to the required data as they are
stored inside the hash table. The caller uses overloaded functions to
get the key and "mapped" values from this pointer. Such an interface
avoids the need for constructing iterator-like wrapper objects, which is
especially important for compound hash tables such as the future
StringHashMap.

This is a preparation for #5417.

Category:
- Code cleanup

Todo:
- rebase to master after sparse dictionary changes
- test the performance -- I've seen an improvement on string_hash_map test, probably because we don't need temporary iterator objects with this patch.

### CODE REVIEW

**Comment by akuzm** [comment]
This has some conflicts with master due to recent sparse dictionary changes. I'll rebase later.

**Comment by akuzm** [comment]
@amosbird @alexey-milovidov @KochetovNicolai for your consideration

**Review by alexey-milovidov — APPROVED** [review]
Good for me.

**Comment by akuzm** [comment]
```
SELECT URL, count() AS c FROM hits_100m_single GROUP BY URL ORDER BY c DESC LIMIT 10

┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━━━┓
┃ version               ┃ count() ┃   min ┃   med ┃   max ┃ q95-q05 ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━━━┩
│ /home/akuzm/ch-master │      49 │ 7.159 │ 7.347 │ 7.468 │   0.274 │
├───────────────────────┼─────────┼───────┼───────┼───────┼─────────┤
│ /home/akuzm/ch-lookup │      49 │ 7.157 │ 7.372 │ 7.554 │   0.294 │
└───────────────────────┴─────────┴───────┴───────┴───────┴─────────┘
```

This is 49 runs for each version, interleaved 7 by 7 to filter out the systemic error. Doesn't show any meaningful difference above the noise.

**Comment by akuzm** [comment]
```
SELECT MobilePhoneModel, uniq(UserID) AS u FROM hits_100m_single WHERE MobilePhoneModel != '' GROUP BY MobilePhoneModel ORDER BY u DESC LIMIT 10 

┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━━━┓
┃ version               ┃ count() ┃   min ┃   med ┃   max ┃ q95-q05 ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━━━┩
│ /home/akuzm/ch-master │      49 │ 0.658 │ 0.673 │ 0.723 │   0.032 │
├───────────────────────┼─────────┼───────┼───────┼───────┼─────────┤
│ /home/akuzm/ch-lookup │      49 │  0.66 │ 0.674 │ 0.895 │    0.05 │
└───────────────────────┴─────────┴───────┴───────┴───────┴─────────┘
```

Looks like a regression. I forgot to make the functions inline, will check again.

File: dbms/src/Common/HashTable/ClearableHashMap.h [hunk 1]
```diff
@@ -15,10 +15,10 @@ struct ClearableHashMapCell : public ClearableHashTableCell<Key, HashMapCell<Key
 };
 
 template<typename Key, typename Mapped, typename Hash>
-auto lookupResultGetKey(ClearableHashMapCell<Key, Mapped, Hash> * cell) { return &cell->getFirst(); }
+ALWAYS_INLINE inline auto lookupResultGetKey(ClearableHashMapCell<Key, Mapped, Hash> * cell) { return &cell->getFirst(); }
```
Comment (amosbird) [inline dbms/src/Common/HashTable/ClearableHashMap.h]: template functions don't need `inline`

Comment (akuzm) [reply dbms/src/Common/HashTable/ClearableHashMap.h]: Without `inline`, gcc-9 gives me warnings like these:

```
/home/akuzm/ch4/ch/dbms/src/Common/HashTable/HashTable.h:211:20: warning: always_inline function might not be inlinable [-Wattributes]
  211 | ALWAYS_INLINE auto lookupResultGetKey(HashTableCell<Key, Hash, State> * cell)
      |                    ^~~~~~~~~~~~~~~~~~
```

Not sure if it's a broken warning, or there is some special rule for partial specialization.

Comment (amosbird) [reply dbms/src/Common/HashTable/ClearableHashMap.h]: yeah, specializations need inline. Not sure about partial ones

Comment (amosbird) [reply dbms/src/Common/HashTable/ClearableHashMap.h]: But templates and partial templates should be exempt from ODR, which means `inline` is not needed.

**Comment by akuzm** [comment]
Looks better with inlining:

```
SELECT URL, count() AS c FROM hits_100m_single GROUP BY URL ORDER BY c DESC LIMIT 10
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━━━┓
┃ version               ┃ count() ┃   min ┃   med ┃   max ┃ q95-q05 ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━━━┩
│ /home/akuzm/ch-master │     119 │ 7.059 │ 7.309 │ 7.448 │   0.244 │
├───────────────────────┼─────────┼───────┼───────┼───────┼─────────┤
│ /home/akuzm/ch-inline │     119 │ 7.068 │   7.3 │  7.51 │   0.249 │
├───────────────────────┼─────────┼───────┼───────┼───────┼─────────┤
│ /home/akuzm/ch-lookup │     119 │ 7.112 │ 7.297 │  7.53 │    0.24 │
└───────────────────────┴─────────┴───────┴───────┴───────┴─────────┘


SELECT MobilePhoneModel, uniq(UserID) AS u FROM hits_100m_single WHERE MobilePhoneModel != '' GROUP BY MobilePhoneModel ORDER BY u DESC LIMIT 10 
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━━━┓
┃ version               ┃ count() ┃   min ┃   med ┃   max ┃ q95-q05 ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━━━┩
│ /home/akuzm/ch-master │     119 │  0.66 │ 0.674 │ 0.755 │   0.047 │
├───────────────────────┼─────────┼───────┼───────┼───────┼─────────┤
│ /home/akuzm/ch-inline │     119 │ 0.656 │ 0.672 │ 0.725 │   0.044 │
├───────────────────────┼─────────┼───────┼───────┼───────┼─────────┤
│ /home/akuzm/ch-lookup │     119 │  0.66 │ 0.673 │ 0.753 │   0.046 │
└───────────────────────┴─────────┴───────┴───────┴───────┴─────────┘

**Pull request merged**